functional_benchmark_test.go

```diff
@@ -0,0 +1,727 @@
+}
+
+func BenchmarkJsArrayMergeSort10000(b *testing.B) {
+	benchmarkJsArraySort(10000, "mergeSort(testSlice);", jsMergeSort, b)
+}
+
+func BenchmarkJsArrayHeapSort10000(b *testing.B) {
+	benchmarkJsArraySort(10000, "heapSort(testSlice);", jsHeapSort, b)
+}
+
+func BenchmarkCryptoAES(b *testing.B) {
+	for i := 0; i < b.N; i++ {
+		vm := New()
+		vm.Run(jsCryptoAES)
+	}
+}
+
+func testGoSliceSort(sortFuncCall string, sortCode string, t *testing.T) {
+	tt(t, func() {
+		test, vm := test()
+
+		// inject quicksort code
+		_, err := vm.Run(sortCode)
+		is(err, nil)
+
+		testSlice := []int{5, 3, 2, 4, 1}
+		vm.Set("testSlice", testSlice)
+		_, err = vm.Run(sortFuncCall)
+		is(err, nil)
+
+		is(test(`testSlice[0]`).export(), 1)
+		is(test(`testSlice[1]`).export(), 2)
+		is(test(`testSlice[2]`).export(), 3)
+		is(test(`testSlice[3]`).export(), 4)
+		is(test(`testSlice[4]`).export(), 5)
+
+		is(testSlice[0], 1)
+		is(testSlice[1], 2)
+		is(testSlice[2], 3)
+		is(testSlice[3], 4)
+		is(testSlice[4], 5)
+	})
+}
+
+func testJsArraySort(sortFuncCall string, sortCode string, t *testing.T) {
+	tt(t, func() {
+		test, vm := test()
+
+		// inject quicksort code
+		vm.Run(sortCode)
+
+		vm.Run("var testSlice = [5, 3, 2, 4, 1];")
+		_, err := vm.Run(sortFuncCall)
+		is(err, nil)
+
+		is(test(`testSlice[0]`).export(), 1)
+		is(test(`testSlice[1]`).export(), 2)
+		is(test(`testSlice[2]`).export(), 3)
+		is(test(`testSlice[3]`).export(), 4)
+		is(test(`testSlice[4]`).export(), 5)
+	})
+}
+
+func benchmarkGoSliceSort(size int, sortFuncCall string, sortCode string, b *testing.B) {
+	// generate arbitrary slice of 'size'
+	testSlice := make([]int, size)
+	for i := 0; i < size; i++ {
+		testSlice[i] = rand.Int()
+	}
+
+	vm := New()
+
+	// inject the sorting code
+	vm.Run(sortCode)
+
+	// Reset timer - everything until this point may have taken a long time
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		vm.Run(sortFuncCall)
+	}
+}
+
+func benchmarkJsArraySort(size int, sortFuncCall string, sortCode string, b *testing.B) {
+	// generate arbitrary slice of 'size'
+	testSlice := make([]string, size)
+	for i := 0; i < size; i++ {
+		testSlice[i] = fmt.Sprintf("%d", rand.Int())
+	}
+
+	jsArrayString := "[" + strings.Join(testSlice, ",") + "]"
+
+	vm := New()
+
+	// inject the test array
+	vm.Run("testSlice = " + jsArrayString)
+
+	// inject the sorting code
+	vm.Run(sortCode)
+
+	// Reset timer - everything until this point may have taken a long time
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		vm.Run(sortFuncCall)
+	}
+}
+
+
+
+
+/**********************************************************************************************************************/
+// Appendix - all the Javascript algorithm code constants
+
+const jsQuickSort = `
+function quickSort(arr, left, right){
+   var len = arr.length,
+   pivot,
+   partitionIndex;
+
+
+  if(left < right){
+    pivot = right;
+    partitionIndex = partition(arr, pivot, left, right);
+
+   //sort left and right
```

Comment:
Also space after //

The comment about style refers to a few lines up but can't edit on a mobile 😋